### ISSUE TITLE
Ability to generate waves

### ISSUE THREAD

**Issue by chrisnorman7** [synthizer/synthizer#17]
Hi,
I've put this as an issue, because I *imagine* the capability is already there:

It would be great if we could generate waves with Synthizer. I'd particularly like sine waves, although I suppose any types would do.

Is this something that is possible / on the roadmap? I'll not complain if it's not.

Thanks for all your hard work.

**Comment by camlorn** [comment]
It's not really on the roadmap, and it's also not already there, but it's easy enough to implement at some point and is generally the kind of thing I want to implement.  SO let's scope it because this is actually a much bigger area than it seems like it is.

What do you want to do with these waves, synthesized beeps?  Chiptunes?

If you know what user stories are, I'm going to propose that you gather/come up with some user stories, but I'm open to scoping this out any way that works for you.

**Comment by chrisnorman7** [comment]
Honestly, I hadn't thought all that far.

In the immediate future, what I want to do is generate waves to make like
an organ setup using the harmonic series. So I suppose I want a wave
generator or something that I can set the type of wave, and the frequency
on.

Beyond that I hadn't thought honestly. Anything you can give me would be
cool though.

Thanks again for all your hard work. I love this library!

Take care,

Chris Norman

**Comment by ultrasound1372** [comment]
Perhaps some standard chip waves, though you can leave out LFSR noise. And of course they'd have to be bandlimited. And if you allow squares and sawtooths, please have the bandlimiting dynamic enough so a low frequency doesn't allow me to clearly hear the top harmonic...

**Comment by camlorn** [comment]
I'm going to just copy WebAudio, which can handle any wave you could ever want and then some.  Consult their spec for PeriodicWave if you want to know what I mean by that, though we'll be adding some helpers, i.e. `syz_periodicWaveConfigureSquare(wave, num_harmonics)`.  Initial implementation will just evaluate it as a continuous function, but I'll eventually throw sinc interpolation on it and we can call it a day.

**Comment by Tetsujinfr** [comment]
Not sure if that helps, but there is a good python sinewave generator library named `pysinewave`.
It does generate sinewaves, but you can also change its pitch continuously and there are no audio cracks. 
fyi.

**Comment by ahicks92** [comment]
This is a C++ library.  Integrating Python libraries into it won't happen, but thanks for trying to help.

Ultimately this issue is pretty easy, I just haven't had the time.

Speed won't be a problem.  If it's  a problem in that library it's because they did something wrong, probably trying to generate the waves in pure Python instead of using Cython or a C extension, or perhaps using the naive method whereby you call `sin()` over and over on every sample.

**Comment by chrisnorman7** [comment]
OK, thank you.

Take care,

Chris Norman

**Comment by ahicks92** [comment]
Over a year after I promised to do this, it's about to land on master.  We now offer an interface which is capable of representing basically all chiptune-style waveforms by directly summing sine waves in a fashion which sacrifices slight accuracy in order to be very fast on modern microprocessors.  Currently, it also provides helpers for square/sawtooth/triangle, though I did these by listening to samples and comparing.  If they don't sound right to you let me know and I will work on it.
